kasa/protocol.py

```diff
@@ -38,41 +47,69 @@ async def query(host: str, request: Union[str, Dict], retry_count: int = 3) -> D
         :param retry_count: how many retries to do in case of failure
         :return: response dict
         """
+        if not self.query_lock:
+            self.query_lock = asyncio.Lock()
+
         if isinstance(request, dict):
             request = json.dumps(request)
 
+        assert isinstance(request, str)
+
         timeout = TPLinkSmartHomeProtocol.DEFAULT_TIMEOUT
-        writer = None
+
+        async with self.query_lock:
+            return await self._query(request, retry_count, timeout)
+
+    async def _connect(self, timeout: int) -> bool:
+        """Try to connect or reconnect to the device."""
+        if self.writer:
+            return True
+        with contextlib.suppress(Exception):
+            task = asyncio.open_connection(
+                self.host, TPLinkSmartHomeProtocol.DEFAULT_PORT
+            )
+            self.reader, self.writer = await asyncio.wait_for(task, timeout=timeout)
+            return True
+        return False
+
+    async def _execute_query(self, request: str) -> Dict:
+        """Execute a query on the device and wait for the response."""
+        assert self.writer is not None
+        assert self.reader is not None
+        _LOGGER.debug("> (%i) %s", len(request), request)
+        self.writer.write(TPLinkSmartHomeProtocol.encrypt(request))
+        await self.writer.drain()
+        packed_block_size = await self.reader.readexactly(self.BLOCK_SIZE)
+        length = struct.unpack(">I", packed_block_size)[0]
+        buffer = await self.reader.readexactly(length)
+        response = TPLinkSmartHomeProtocol.decrypt(buffer)
+        json_payload = json.loads(response)
+        _LOGGER.debug("< (%i) %s", len(response), pf(json_payload))
+        return json_payload
+
+    async def _close(self):
+        """Close the connection."""
+        if self.writer:
+            self.writer.close()
+            with contextlib.suppress(Exception):
+                await self.writer.wait_closed()
+        self.writer = None
+        self.reader = None
```

Comment:
This shares mostly the same code as `__del__` (besides awaiting for closing), consolidate?

Reply:
I made it a bit more dry, but can't do so much since we can't await in `__del__`